### PR TITLE
Fix SplitV op

### PR DESCRIPTION
**Pull request by kanvi-nervana**
- Add invalid argument errors if size_splits != length

### CODE REVIEW

File: src/ngraph_builder.cc [hunk 1]
```diff
@@ -3332,32 +3332,45 @@ static Status TranslateSplitVOp(
...
+  // Find out the total length of the splits
+  for (int i = 0; i < lengths.size(); ++i) {
+    if (lengths[i] != -1) {
+      length += lengths[i];
+    } else {
+      if(has_one_neg) {
+        return errors::InvalidArgument(
+          "size_splits can only have one -1");
+      } else {
+        idx = i;
```
Comment (shresthamalik) [inline src/ngraph_builder.cc]: Need to set has_one_neg true, in else. May be has_only_one_neg would be more explicable?

File: src/ngraph_builder.cc [hunk 2]
```diff
@@ -3332,32 +3332,45 @@ static Status TranslateSplitVOp(
...
+      }
+    }
+  }
+
+  if(idx > 0) {
     lengths[idx] = shape[split_dim] - length;
   }
 
+  if(length != shape[split_dim] || length != shape[split_dim] - lengths[idx]) {
```
Comment (shresthamalik) [inline src/ngraph_builder.cc]: Please add two separate conditions. ```(length!=shape[split_dim)``` is a valid error condition when ```idx==-1```

File: src/ngraph_builder.cc [hunk 3]
```diff
@@ -3332,32 +3332,45 @@ static Status TranslateSplitVOp(
...
-  // Size splits must sum to the dimension of value along split_dim
-  if (idx > 0) {
+  bool has_one_neg = false;
+  
+  // length: Length of size_splits
+  int length = 0;
+  int idx = -1;
+
+  // Find out the total length of the splits
+  for (int i = 0; i < lengths.size(); ++i) {
```
Comment (sayantan-nervana) [inline src/ngraph_builder.cc]: The comment before this for loop says: `Find out the total length of the splits`, but from what I understand it is locating the one negative value if present.

Also, I think it was checked last time, but can we recheck what happens if we pass if a neg number other than -1? Are we relying on TF complaining about it during construction?

Comment (kanvi-nervana) [reply src/ngraph_builder.cc]: TF complains if we specify any other neg number other than -1

File: src/ngraph_builder.cc [hunk 4]
```diff
@@ -3332,32 +3332,46 @@ static Status TranslateSplitVOp(
...
-  // Size splits must sum to the dimension of value along split_dim
-  if (idx > 0) {
+  bool has_one_neg = false;
+  
+  // length: Length of size_splits
+  int length = 0;
+  int idx = -1;
+
+  // Find out the total length of the splits and locate it's index
+  for (int i = 0; i < lengths.size(); ++i) {
```
Comment (sayantan-nervana) [inline src/ngraph_builder.cc]: Minor:

`locate it's index` -> `locate it's negative (-1) index, if any`

File: src/ngraph_builder.cc [hunk 5]
```diff
@@ -3332,32 +3332,46 @@ static Status TranslateSplitVOp(
...
+        has_one_neg = true;
+      }
+    }
+  }
+
+  if(idx > 0) {
     lengths[idx] = shape[split_dim] - length;
   }
 
+  if((idx == -1 && length != shape[split_dim]) || (idx > 0 && lengths[idx] <= 0)) {
```
Comment (sayantan-nervana) [inline src/ngraph_builder.cc]: Minor readability comment:
instead of `idx == -1` we can use `!has_one_neg`, which might be more readable
Similarly for `idx > 0` we can use `has_one_neg`
So: `(!has_one_neg && length != shape[split_dim]) || (has_one_neg && lengths[idx] <= 0)`

Possible corner case:
I am wondering about the `<= 0` part.. Can a split have `0` length? We could have a test. If `0` size splits are acceptable, then we need to make that `< 0` and not `<= 0`

File: src/ngraph_builder.cc [hunk 6]
```diff
@@ -3332,32 +3332,46 @@ static Status TranslateSplitVOp(
...
+          return errors::InvalidArgument(
+          "The length of size_splits must sum to the dimension of value along split_dim");
```
Comment (sayantan-nervana) [inline src/ngraph_builder.cc]: `split_dim` is an internal local variable. We should use `axis`, which is the TF name for this variable.
`dimension of value` -> did you mean `value of dimension`?

Also you could print out the value of `axis`

Ideally, the 2 error cases should get different error messages. Maybe we can split them into 2 `if`s, instead of `or`ing the error condition `((idx == -1 && length != shape[split_dim]) || (idx > 0 && lengths[idx] <= 0))`
